cf/data/data.py

```diff
@@ -13325,3 +13154,146 @@ def _broadcast(a, shape):
     tile = shape[0 : len(shape) - len(a_shape)] + tuple(tile[::-1])
 
     return np.tile(a, tile)
+
+
+def _where_broadcastable(data, x, name):
+    """Check broadcastability for `where` assignments.
+
+    Raises an exception if the result of broadcasting *data* and *x*
+    together does not have the same shape as *data*.
+
+    .. versionadded:: TODODASK
+
+    .. seealso:: `where`
+
+    :Parameters:
+
+        data, x: `Data`
+            The arrays to compare.
+
+        name: `str`
+            A name for *x* that is used in any exception error
+            message.
+
+    :Returns:
+
+        `bool`
+             If *x* is acceptably broadcastable to *data* then `True`
+             is returned, otherwise a `ValueError` is raised.
+
+    """
+    ndim_x = x.ndim
+    if not ndim_x:
+        return True
+
+    ndim_data = data.ndim
+    if ndim_x > ndim_data:
+        raise ValueError(
+            f"where: Broadcasting the {name!r} parameter with {ndim_x} "
+            f"dimensions would change the shape of the data with "
+            f"{ndim_data} dimensions"
+        )
+
+    shape_x = x.shape
+    shape_data = data.shape
+    for n, m in zip(shape_x[::-1], shape_data[::-1]):
+        if n != m and n != 1:
```

Comment:
Potentially my thought as follows would be premature optimisation, so maybe if anything it could be a tentative `TODODASK` comment, but could this logic be refactored to avoid the need to calculate the `shape` of the data in the cases across where `x` has only size-1 dimensions? In other words, to essentially change the short-circuiting on the `if` check here to:

```suggestion
        if n != 1 n != m:
```

in case `x` is trivial and `data` is not, notably if it is large so expensive to get its `shape`? Would that even be worthwhile or could it e.g. be a niche case with general usage of the `where` method (not sure right now, thinking aloud here)?

Reply:
Can we talk about this one on out next call?

Reply:
@davidhassell, sure. I'll add it to my (sprawling) questions and thoughts list.